In [373]:
import pandas as pd
import numpy as np

In [374]:
data_frame = pd.read_json('https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json')

In [375]:
data_frame.head(10)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
5  0013-MHZWF    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
6  0013-SMEOE    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
7  0014-BMAQU    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
8  0015-UOCOJ    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
9  0016-QLJIS    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
5   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
8   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
9  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   
5  {'InternetService': 'DSL', 'OnlineSecurity': '...   
6  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7  {'InternetService': 'Fiber optic', 'OnlineSecu...   
8  {'InternetService': 'DSL', 'OnlineSecurity': '...   
9  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...  
5  {'Contract': 'Month-to-month', 'PaperlessBilli...  
6  {'Contract': 'Two year', 'PaperlessBilling': '...  
7  {'Contract': 'Two year', 'PaperlessBilling': '...  
8  {'Contract': 'Month-to-month', 'PaperlessBilli...  
9  {'Contract': 'Two year', 'PaperlessBilling': '...

In [376]:
len(data_frame)

7267

In [377]:
type(data_frame)

pandas.core.frame.DataFrame

In [378]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [379]:
data_frame.dtypes

customerID    object
Churn         object
customer      object
phone         object
internet      object
account       object
dtype: object

In [380]:
tipos_de_dados = pd.DataFrame(data_frame.dtypes, columns = ['Tipos de Dados'])

In [381]:
tipos_de_dados.columns.name = 'Variáveis'

In [382]:
tipos_de_dados

Variáveis  Tipos de Dados
customerID         object
Churn              object
customer           object
phone              object
internet           object
account            object

In [383]:
data_frame.shape

(7267, 6)

In [384]:
print(f'Número de Variáveis: {data_frame.shape[1]} \nNúmero de IDs cadastrados: {data_frame.shape[0]}')


Número de Variáveis: 6 
Número de IDs cadastrados: 7267


In [385]:
data_frame.head(5)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [386]:
yes = ['Yes']
selecao_yes = data_frame['Churn'].isin(yes)
churn_yes = data_frame[selecao_yes]

In [387]:
churn_yes.head(5)

customerID Churn                                           customer  \
2   0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3   0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4   0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
18  0022-TCJCI   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
19  0023-HGHWL   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   

                                                phone  \
2      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4      {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
18     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
19  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                             internet  \
2   {'InternetService': 'Fiber optic', 'OnlineSecu...   
3   {'InternetService': 'Fiber optic', 'OnlineSecu...   
4   {'InternetService': 'Fiber optic', 'OnlineSecu...   
18  {'InternetService': 'DSL', 'OnlineSecurity': '...   
19  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                              account  
2   {'Contract': 'Month-to-month', 'PaperlessBilli...  
3   {'Contract': 'Month-to-month', 'PaperlessBilli...  
4   {'Contract': 'Month-to-month', 'PaperlessBilli...  
18  {'Contract': 'One year', 'PaperlessBilling': '...  
19  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [388]:
no = ['No']
selecao_no = data_frame['Churn'].isin(no)
churn_no = data_frame[selecao_no]

In [389]:
churn_no.head(5)

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
5  0013-MHZWF    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
6  0013-SMEOE    No  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
7  0014-BMAQU    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
5   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
6   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
5  {'InternetService': 'DSL', 'OnlineSecurity': '...   
6  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
5  {'Contract': 'Month-to-month', 'PaperlessBilli...  
6  {'Contract': 'Two year', 'PaperlessBilling': '...  
7  {'Contract': 'Two year', 'PaperlessBilling': '...

In [390]:
data_frame['Churn'].drop_duplicates()

0      No
2     Yes
30       
Name: Churn, dtype: object

In [391]:
vazio = ['']
selecao_vazio = data_frame['Churn'].isin(vazio)
churn_vazio = data_frame[selecao_vazio]


In [392]:
churn_vazio.head(5)

customerID Churn                                           customer  \
30   0047-ZHDTW        {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
75   0120-YZLQA        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
96   0154-QYHJU        {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
98   0162-RZGMZ        {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
175  0274-VVQOQ        {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   

                                               phone  \
30   {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
75    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
96    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
98    {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
175  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   

                                              internet  \
30   {'InternetService': 'Fiber optic', 'OnlineSecu...   
75   {'InternetService': 'No', 'OnlineSecurity': 'N...   
96   {'InternetService': 'DSL', 'OnlineSecurity': '...   
98   {'InternetService': 'DSL', 'OnlineSecurity': '...   
175  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                               account  
30   {'Contract': 'Month-to-month', 'PaperlessBilli...  
75   {'Contract': 'Two year', 'PaperlessBilling': '...  
96   {'Contract': 'One year', 'PaperlessBilling': '...  
98   {'Contract': 'Month-to-month', 'PaperlessBilli...  
175  {'Contract': 'One year', 'PaperlessBilling': '...

In [393]:
soma_churn = len(churn_yes) + len(churn_no) + len(churn_vazio)

In [394]:
print(f'Total de clientes: {soma_churn}\nDesistência de clientes: {len(churn_yes)}\nPermanência de clientes: {len(churn_no)}\nCampo Vazio: {len(churn_vazio)}')

Total de clientes: 7267
Desistência de clientes: 1869
Permanência de clientes: 5174
Campo Vazio: 224


# customer	

In [395]:
customer = pd.read_csv('dados_pessoais.csv')

In [396]:
customer.head(5)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes
0  0002-ORFBO          No  Female            0          Yes             Yes
1  0003-MKNFE          No    Male            0           No              No
2  0004-TLHLJ         Yes    Male            0           No              No
3  0011-IGKFF         Yes    Male            1          Yes              No
4  0013-EXCHZ         Yes  Female            1          Yes              No

In [397]:
customer = customer.rename(columns = {'Churn': 'desistencia'})

In [398]:
customer.head(5)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes
0  0002-ORFBO          No  Female            0          Yes             Yes
1  0003-MKNFE          No    Male            0           No              No
2  0004-TLHLJ         Yes    Male            0           No              No
3  0011-IGKFF         Yes    Male            1          Yes              No
4  0013-EXCHZ         Yes  Female            1          Yes              No

In [399]:
customer.to_csv('dados_pessoais.csv', index=False)

#####################################################
INFORMAÇÕES PESSOAIS DO CLIENTE:
#####################################################


Sexo Masculino: 3675

Sexo Feminino: 3592

Total: 7267

--------------------------------------------------


Idade >= 65: 1182

Idade < 65: 6085

Total: 7267

--------------------------------------------------


Cliente tem parceiro: 3518

Cliente não tem parceiro: 3749

Total: 7267

--------------------------------------------------


Cliente tem dependentes: 2181

Cliente não tem parceiro: 5086

Total: 7267

#####################################################
DESISTÊNCIA DO CLIENTE:
#####################################################

Desistência de clientes: 1869

Permanência de clientes: 5174

Campo Vazio: 224

Total de clientes: 7267

# phone

In [400]:
phone = pd.read_csv('dados_telefonicos.csv')

In [401]:
phone.head(5)

cliente_ID servico_telefone servico_telefone_multiplas_linhas
0  0002-ORFBO              Yes                                No
1  0003-MKNFE              Yes                               Yes
2  0004-TLHLJ              Yes                                No
3  0011-IGKFF              Yes                                No
4  0013-EXCHZ              Yes                                No

##################################################
            INFORMAÇÕES TELEFONE
##################################################

Assinatura de serviço telefônico com uma linha

  Sim: 6560

  Não: 707

Total: 7267

--------------------------------------------------

Assisnatura de mais de uma linha de telefone:

  Sim: 3065

  Não: 3495

  Sem serviço de telefone: 707

Total: 7267

# internet

In [402]:
internet = pd.read_csv('servicos_internet.csv')

In [403]:
internet.head(5)

cliente_ID assinatura_provedor_internet  \
0  0002-ORFBO                          DSL   
1  0003-MKNFE                          DSL   
2  0004-TLHLJ                  Fiber optic   
3  0011-IGKFF                  Fiber optic   
4  0013-EXCHZ                  Fiber optic   

  assinatura_adicional_seguranca_online assinatura_adicional_backup_online  \
0                                    No                                Yes   
1                                    No                                 No   
2                                    No                                 No   
3                                    No                                Yes   
4                                    No                                 No   

  assinatura_adicional_protecao_dispositivo  \
0                                        No   
1                                        No   
2                                       Yes   
3                                       Yes   
4                                        No   

  assinatura_adicional_suporte_tecnico_menos_tempo_espera assinatura_TV_cabo  \
0                                                Yes                     Yes   
1                                                 No                      No   
2                                                 No                      No   
3                                                 No                     Yes   
4                                                Yes                     Yes   

  assinatura_streaming_filmes  
0                          No  
1                         Yes  
2                          No  
3                         Yes  
4                          No

##################################################
INFORMAÇÕES SOBRE INTERNET
##################################################

Assinatura de um provedor internet:

  DSL:2488

  Fibra óptica: 3198

  Sem Provedor: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de segurança online:

  Sim: 2078

  Não: 3608

  Sem serviço de internet: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de backup online:

  Sim: 2504

  Não: 3182

  Sem serviço de internet: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de proteção no dispositivo:

  Sim: 2491

  Não: 3195

  Sem serviço de internet: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de suporte técnico, menos tempo de espera:

  Sim: 2104

  Não: 3582

  Sem serviço de internet: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de TV a cabo:

  Sim: 2790

  Não: 2896

  Sem serviço de internet: 1581

Total: 7267

--------------------------------------------------

Assinatura adicional de streaming de filmes:

  Sim: 2816

  Não: 2870

  Sem serviço de internet: 1581

Total: 7267


# account

In [404]:
account = pd.read_csv('conta.csv')

In [405]:
account.head(5)

cliente_ID        contrato fatura_online   forma_pagamento  gastos_mensais  \
0  0002-ORFBO        One year           Yes      Mailed check            65.6   
1  0003-MKNFE  Month-to-month            No      Mailed check            59.9   
2  0004-TLHLJ  Month-to-month           Yes  Electronic check            73.9   
3  0011-IGKFF  Month-to-month           Yes  Electronic check            98.0   
4  0013-EXCHZ  Month-to-month           Yes      Mailed check            83.9   

  gasto_total  
0       593.3  
1       542.4  
2      280.85  
3     1237.85  
4       267.4

##################################################
INFORMAÇÕES SOBRE CONTA
##################################################

Contratos:

  Contrato de 1 Ano: 1519

  Contrato de 2 Anos: 1743

  Contrato de Mês a Mês: 4005

Total: 7267

--------------------------------------------------

Receber fatura online:

  Sim: 4311

  Não: 2956

Total: 7267

--------------------------------------------------

Forma de Pagamento:

  Cheque Enviado: 1665

  Cheque Eletrônico: 2445

  Cartão de Crédito(automático): 1568

  Transferência Bancaria(automático): 1589

Total: 7267



In [406]:
dados = pd.read_csv('churn_de_clientes.csv')

In [407]:
dados.head(5)

cliente_ID Churn  genero  idade_>=_65 tem_parceiro tem_dependentes  \
0  0002-ORFBO    No  Female            0          Yes             Yes   
1  0003-MKNFE    No    Male            0           No              No   
2  0004-TLHLJ   Yes    Male            0           No              No   
3  0011-IGKFF   Yes    Male            1          Yes              No   
4  0013-EXCHZ   Yes  Female            1          Yes              No   

  assinatura_provedor_internet assinatura_adicional_seguranca_online  \
0                          DSL                                    No   
1                          DSL                                    No   
2                  Fiber optic                                    No   
3                  Fiber optic                                    No   
4                  Fiber optic                                    No   

  assinatura_adicional_backup_online  \
0                                Yes   
1                                 No   
2                                 No   
3                                Yes   
4                                 No   

  assinatura_adicional_protecao_dispositivo  \
0                                        No   
1                                        No   
2                                       Yes   
3                                       Yes   
4                                        No   

  assinatura_adicional_suporte_tecnico_menos_tempo_espera assinatura_TV_cabo  \
0                                                Yes                     Yes   
1                                                 No                      No   
2                                                 No                      No   
3                                                 No                     Yes   
4                                                Yes                     Yes   

  assinatura_streaming_filmes        contrato fatura_online   forma_pagamento  \
0                          No        One year           Yes      Mailed check   
1                         Yes  Month-to-month            No      Mailed check   
2                          No  Month-to-month           Yes  Electronic check   
3                         Yes  Month-to-month           Yes  Electronic check   
4                          No  Month-to-month           Yes      Mailed check   

   gastos_mensais gasto_total  
0            65.6       593.3  
1            59.9       542.4  
2            73.9      280.85  
3            98.0     1237.85  
4            83.9       267.4

In [408]:
dados = dados.rename(columns = {'Churn': 'desistencia'})

In [409]:
dados.head(5)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes  \
0  0002-ORFBO          No  Female            0          Yes             Yes   
1  0003-MKNFE          No    Male            0           No              No   
2  0004-TLHLJ         Yes    Male            0           No              No   
3  0011-IGKFF         Yes    Male            1          Yes              No   
4  0013-EXCHZ         Yes  Female            1          Yes              No   

  assinatura_provedor_internet assinatura_adicional_seguranca_online  \
0                          DSL                                    No   
1                          DSL                                    No   
2                  Fiber optic                                    No   
3                  Fiber optic                                    No   
4                  Fiber optic                                    No   

  assinatura_adicional_backup_online  \
0                                Yes   
1                                 No   
2                                 No   
3                                Yes   
4                                 No   

  assinatura_adicional_protecao_dispositivo  \
0                                        No   
1                                        No   
2                                       Yes   
3                                       Yes   
4                                        No   

  assinatura_adicional_suporte_tecnico_menos_tempo_espera assinatura_TV_cabo  \
0                                                Yes                     Yes   
1                                                 No                      No   
2                                                 No                      No   
3                                                 No                     Yes   
4                                                Yes                     Yes   

  assinatura_streaming_filmes        contrato fatura_online   forma_pagamento  \
0                          No        One year           Yes      Mailed check   
1                         Yes  Month-to-month            No      Mailed check   
2                          No  Month-to-month           Yes  Electronic check   
3                         Yes  Month-to-month           Yes  Electronic check   
4                          No  Month-to-month           Yes      Mailed check   

   gastos_mensais gasto_total  
0            65.6       593.3  
1            59.9       542.4  
2            73.9      280.85  
3            98.0     1237.85  
4            83.9       267.4

In [410]:
dados.isnull()

cliente_ID  desistencia  genero  idade_>=_65  tem_parceiro  \
0          False        False   False        False         False   
1          False        False   False        False         False   
2          False        False   False        False         False   
3          False        False   False        False         False   
4          False        False   False        False         False   
...          ...          ...     ...          ...           ...   
7262       False        False   False        False         False   
7263       False        False   False        False         False   
7264       False        False   False        False         False   
7265       False        False   False        False         False   
7266       False        False   False        False         False   

      tem_dependentes  assinatura_provedor_internet  \
0               False                         False   
1               False                         False   
2               False                         False   
3               False                         False   
4               False                         False   
...               ...                           ...   
7262            False                         False   
7263            False                         False   
7264            False                         False   
7265            False                         False   
7266            False                         False   

      assinatura_adicional_seguranca_online  \
0                                     False   
1                                     False   
2                                     False   
3                                     False   
4                                     False   
...                                     ...   
7262                                  False   
7263                                  False   
7264                                  False   
7265                                  False   
7266                                  False   

      assinatura_adicional_backup_online  \
0                                  False   
1                                  False   
2                                  False   
3                                  False   
4                                  False   
...                                  ...   
7262                               False   
7263                               False   
7264                               False   
7265                               False   
7266                               False   

      assinatura_adicional_protecao_dispositivo  \
0                                         False   
1                                         False   
2                                         False   
3                                         False   
4                                         False   
...                                         ...   
7262                                      False   
7263                                      False   
7264                                      False   
7265                                      False   
7266                                      False   

      assinatura_adicional_suporte_tecnico_menos_tempo_espera  \
0                                                 False         
1                                                 False         
2                                                 False         
3                                                 False         
4                                                 False         
...                                                 ...         
7262                                              False         
7263                                              False         
7264                                              False         
7265                                              False         
7266                                              False         

      assinatura_TV_cabo  assinatura_str

In [411]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 18 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   cliente_ID                                               7267 non-null   object 
 1   desistencia                                              7043 non-null   object 
 2   genero                                                   7267 non-null   object 
 3   idade_>=_65                                              7267 non-null   int64  
 4   tem_parceiro                                             7267 non-null   object 
 5   tem_dependentes                                          7267 non-null   object 
 6   assinatura_provedor_internet                             7267 non-null   object 
 7   assinatura_adicional_seguranca_online                    7267 non-null   object 
 8   assinatura_adicional_backup_

In [412]:
no_service = ['No internet service']

In [413]:
selecao = dados['assinatura_adicional_seguranca_online'].isin(no_service)

In [414]:
no_service = dados[selecao]

In [415]:
no_service

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro  \
20    0023-UYUPN          No  Female            1          Yes   
23    0030-FNXPP          No  Female            0           No   
24    0031-PVLZI         Yes  Female            0          Yes   
27    0040-HALCW          No    Male            0          Yes   
28    0042-JVWOJ          No    Male            0           No   
...          ...         ...     ...          ...          ...   
7250  9970-QBCDA          No  Female            0           No   
7252  9972-EWRJS          No  Female            0          Yes   
7256  9975-GPKZU          No    Male            0          Yes   
7257  9975-SKRNR          No    Male            0           No   
7261  9986-BONCE         Yes  Female            0           No   

     tem_dependentes assinatura_provedor_internet  \
20                No                           No   
23                No                           No   
24               Yes                           No   
27               Yes                           No   
28                No                           No   
...              ...                          ...   
7250              No                           No   
7252             Yes                           No   
7256             Yes                           No   
7257              No                           No   
7261              No                           No   

     assinatura_adicional_seguranca_online assinatura_adicional_backup_online  \
20                     No internet service                No internet service   
23                     No internet service                No internet service   
24                     No internet service                No internet service   
27                     No internet service                No internet service   
28                     No internet service                No internet service   
...                                    ...                                ...   
7250                   No internet service                No internet service   
7252                   No internet service                No internet service   
7256                   No internet service                No internet service   
7257                   No internet service                No internet service   
7261                   No internet service                No internet service   

     assinatura_adicional_protecao_dispositivo  \
20                         No internet service   
23                         No internet service   
24                         No internet service   
27                         No internet service   
28                         No internet service   
...                                        ...   
7250                       No internet service   
7252                       No internet service   
7256                       No internet service   
7257                       No internet service   
7261                       No internet service   

     assinatura_adicional_suporte_tecnico_menos_tempo_espera  \
20                                  No internet service        
23                                  No internet service        
24                                  No internet service        
27                                  No internet service        
28                                  No internet service        
...                                                 ...        
7250                                No internet service        
7252                                No internet service        
7256                                No internet service        
7257                                No internet service        
7261                                No internet service        

       assinatura_TV_cabo assinatura_streaming_filmes        contrato  \
20    No internet service         No internet service        One year   
23    No internet service         No internet service  Month-to-month   
24    No inter

In [416]:
print(f'{len(no_service)} clientes não tem contrato de nenhum serviço e {len(desistencia_vazio)} tem a coluna desistencia vazia' )

1581 clientes não tem contrato de nenhum serviço e 224 tem a coluna desistencia vazia


# eliminando colunas sem nenhum tipo de serviço ou com a coluna de desistência vazia

## apagar linhas vazias da coluna desistencia

In [417]:
desistencia_vazio = dados[dados['desistencia'].isnull()]

In [418]:
desistencia_vazio

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro  \
30    0047-ZHDTW         NaN  Female            0           No   
75    0120-YZLQA         NaN    Male            0           No   
96    0154-QYHJU         NaN    Male            0           No   
98    0162-RZGMZ         NaN  Female            1           No   
175   0274-VVQOQ         NaN    Male            1          Yes   
...          ...         ...     ...          ...          ...   
7158  9840-GSRFX         NaN  Female            0           No   
7180  9872-RZQQB         NaN  Female            0          Yes   
7211  9920-GNDMB         NaN    Male            0           No   
7239  9955-RVWSC         NaN  Female            0          Yes   
7247  9966-VYRTZ         NaN  Female            0          Yes   

     tem_dependentes assinatura_provedor_internet  \
30                No                  Fiber optic   
75                No                           No   
96                No                          DSL   
98                No                          DSL   
175               No                  Fiber optic   
...              ...                          ...   
7158              No                          DSL   
7180              No                          DSL   
7211              No                  Fiber optic   
7239             Yes                           No   
7247             Yes                           No   

     assinatura_adicional_seguranca_online assinatura_adicional_backup_online  \
30                                     Yes                                 No   
75                     No internet service                No internet service   
96                                     Yes                                Yes   
98                                     Yes                                Yes   
175                                     No                                Yes   
...                                    ...                                ...   
7158                                    No                                Yes   
7180                                   Yes                                 No   
7211                                    No                                 No   
7239                   No internet service                No internet service   
7247                   No internet service                No internet service   

     assinatura_adicional_protecao_dispositivo  \
30                                          No   
75                         No internet service   
96                                          No   
98                                          No   
175                                        Yes   
...                                        ...   
7158                                        No   
7180                                        No   
7211                                        No   
7239                       No internet service   
7247                       No internet service   

     assinatura_adicional_suporte_tecnico_menos_tempo_espera  \
30                                                   No        
75                                  No internet service        
96                                                  Yes        
98                                                  Yes        
175                                                  No        
...                                                 ...        
7158                                                 No        
7180                                                 No        
7211                                                 No        
7239                                No internet service        
7247                                No internet service        

       assinatura_TV_cabo assinatura_streaming_filmes        contrato  \
30                     No                          No  Month-to-month   
75    No internet service         No internet service        Two year   
96            

In [419]:
A = dados.shape[0]
dados.dropna(subset = ['desistencia'], inplace = True)
B = dados.shape[0]
A-B

224

In [420]:
dados[dados['desistencia'].isnull()]

Empty DataFrame
Columns: [cliente_ID, desistencia, genero, idade_>=_65, tem_parceiro, tem_dependentes, assinatura_provedor_internet, assinatura_adicional_seguranca_online, assinatura_adicional_backup_online, assinatura_adicional_protecao_dispositivo, assinatura_adicional_suporte_tecnico_menos_tempo_espera, assinatura_TV_cabo, assinatura_streaming_filmes, contrato, fatura_online, forma_pagamento, gastos_mensais, gasto_total]
Index: []

In [421]:
len(dados)

7043

In [422]:
selecao = dados['assinatura_adicional_seguranca_online'].isin(['No internet service'])

In [423]:
A = dados.shape[0]
dados = dados[~selecao]
B = dados.shape[0]
A - B

1526

In [424]:
dados.head(30)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes  \
0   0002-ORFBO          No  Female            0          Yes             Yes   
1   0003-MKNFE          No    Male            0           No              No   
2   0004-TLHLJ         Yes    Male            0           No              No   
3   0011-IGKFF         Yes    Male            1          Yes              No   
4   0013-EXCHZ         Yes  Female            1          Yes              No   
5   0013-MHZWF          No  Female            0           No             Yes   
6   0013-SMEOE          No  Female            1          Yes              No   
7   0014-BMAQU          No    Male            0          Yes              No   
8   0015-UOCOJ          No  Female            1           No              No   
9   0016-QLJIS          No  Female            0          Yes             Yes   
10  0017-DINOC          No    Male            0           No              No   
11  0017-IUDMW          No  Female            0          Yes             Yes   
12  0018-NYROU          No  Female            0          Yes              No   
13  0019-EFAEP          No  Female            0           No              No   
14  0019-GFNTW          No  Female            0           No              No   
15  0020-INWCK          No  Female            0          Yes             Yes   
16  0020-JDNXP          No  Female            0          Yes             Yes   
17  0021-IKXGC          No  Female            1           No              No   
18  0022-TCJCI         Yes    Male            1           No              No   
19  0023-HGHWL         Yes    Male            1           No              No   
21  0023-XUOPT         Yes  Female            0          Yes              No   
22  0027-KWYKW          No  Female            0          Yes             Yes   
25  0032-PGELS         Yes  Female            0          Yes             Yes   
26  0036-IHMOT          No  Female            0          Yes             Yes   
31  0048-LUMLS          No    Male            0          Yes             Yes   
33  0052-DCKON          No    Male            0          Yes              No   
35  0056-EPFBG          No    Male            0          Yes             Yes   
37  0058-EVZWM          No  Female            0          Yes              No   
38  0060-FUALY          No  Female            0          Yes              No   
40  0064-YIJGF          No    Male            0          Yes             Yes   

   assinatura_provedor_internet assinatura_adicional_seguranca_online  \
0                           DSL                                    No   
1                           DSL                                    No   
2                   Fiber optic                                    No   
3                   Fiber optic                                    No   
4                   Fiber optic                                    No   
5                           DSL                                    No   
6                   Fiber optic                                   Yes   
7                   Fiber optic                                   Yes   
8                           DSL                                   Yes   
9                           DSL                                   Yes   
10                          DSL                                   Yes   
11                  Fiber optic                                   Yes   
12                  Fiber optic                                    No   
13                  Fiber optic                                   Yes   
14                          DSL                                   Yes   
15                  Fiber optic                                    No   
16                          DSL                                   Yes   
17                  Fiber optic                                    No   
18                          DSL                                   Yes   
19                          DSL                                

In [425]:
type(dados)

pandas.core.frame.DataFrame

In [426]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5517 entries, 0 to 7266
Data columns (total 18 columns):
 #   Column                                                   Non-Null Count  Dtype  
---  ------                                                   --------------  -----  
 0   cliente_ID                                               5517 non-null   object 
 1   desistencia                                              5517 non-null   object 
 2   genero                                                   5517 non-null   object 
 3   idade_>=_65                                              5517 non-null   int64  
 4   tem_parceiro                                             5517 non-null   object 
 5   tem_dependentes                                          5517 non-null   object 
 6   assinatura_provedor_internet                             5517 non-null   object 
 7   assinatura_adicional_seguranca_online                    5517 non-null   object 
 8   assinatura_adicional_backup_

# Atualização das Informações

In [427]:
dados.dtypes

cliente_ID                                                  object
desistencia                                                 object
genero                                                      object
idade_>=_65                                                  int64
tem_parceiro                                                object
tem_dependentes                                             object
assinatura_provedor_internet                                object
assinatura_adicional_seguranca_online                       object
assinatura_adicional_backup_online                          object
assinatura_adicional_protecao_dispositivo                   object
assinatura_adicional_suporte_tecnico_menos_tempo_espera     object
assinatura_TV_cabo                                          object
assinatura_streaming_filmes                                 object
contrato                                                    object
fatura_online                                               ob

In [428]:
tipos_de_dados = pd.DataFrame(dados.dtypes, columns = ['Tipos de Dados'])
tipos_de_dados.columns.name = 'Variáveis'
tipos_de_dados

Variáveis                                          Tipos de Dados
cliente_ID                                                 object
desistencia                                                object
genero                                                     object
idade_>=_65                                                 int64
tem_parceiro                                               object
tem_dependentes                                            object
assinatura_provedor_internet                               object
assinatura_adicional_seguranca_online                      object
assinatura_adicional_backup_online                         object
assinatura_adicional_protecao_dispositivo                  object
assinatura_adicional_suporte_tecnico_menos_temp...         object
assinatura_TV_cabo                                         object
assinatura_streaming_filmes                                object
contrato                                                   object
fatura_online                                              object
forma_pagamento                                            object
gastos_mensais                                            float64
gasto_total                                                object

In [429]:
print(f'A base de dados apresenta {dados.shape[0]} e {dados.shape[1]} variáveis')

A base de dados apresenta 5517 e 18 variáveis


# acrescentar coluna gastos_diarios

In [430]:
dados.head(5)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes  \
0  0002-ORFBO          No  Female            0          Yes             Yes   
1  0003-MKNFE          No    Male            0           No              No   
2  0004-TLHLJ         Yes    Male            0           No              No   
3  0011-IGKFF         Yes    Male            1          Yes              No   
4  0013-EXCHZ         Yes  Female            1          Yes              No   

  assinatura_provedor_internet assinatura_adicional_seguranca_online  \
0                          DSL                                    No   
1                          DSL                                    No   
2                  Fiber optic                                    No   
3                  Fiber optic                                    No   
4                  Fiber optic                                    No   

  assinatura_adicional_backup_online  \
0                                Yes   
1                                 No   
2                                 No   
3                                Yes   
4                                 No   

  assinatura_adicional_protecao_dispositivo  \
0                                        No   
1                                        No   
2                                       Yes   
3                                       Yes   
4                                        No   

  assinatura_adicional_suporte_tecnico_menos_tempo_espera assinatura_TV_cabo  \
0                                                Yes                     Yes   
1                                                 No                      No   
2                                                 No                      No   
3                                                 No                     Yes   
4                                                Yes                     Yes   

  assinatura_streaming_filmes        contrato fatura_online   forma_pagamento  \
0                          No        One year           Yes      Mailed check   
1                         Yes  Month-to-month            No      Mailed check   
2                          No  Month-to-month           Yes  Electronic check   
3                         Yes  Month-to-month           Yes  Electronic check   
4                          No  Month-to-month           Yes      Mailed check   

   gastos_mensais gasto_total  
0            65.6       593.3  
1            59.9       542.4  
2            73.9      280.85  
3            98.0     1237.85  
4            83.9       267.4

In [431]:
dados.insert(17,'gastos_diarios',((dados['gastos_mensais'] / 30).round(2)))

In [432]:
dados.head(5)

cliente_ID desistencia  genero  idade_>=_65 tem_parceiro tem_dependentes  \
0  0002-ORFBO          No  Female            0          Yes             Yes   
1  0003-MKNFE          No    Male            0           No              No   
2  0004-TLHLJ         Yes    Male            0           No              No   
3  0011-IGKFF         Yes    Male            1          Yes              No   
4  0013-EXCHZ         Yes  Female            1          Yes              No   

  assinatura_provedor_internet assinatura_adicional_seguranca_online  \
0                          DSL                                    No   
1                          DSL                                    No   
2                  Fiber optic                                    No   
3                  Fiber optic                                    No   
4                  Fiber optic                                    No   

  assinatura_adicional_backup_online  \
0                                Yes   
1                                 No   
2                                 No   
3                                Yes   
4                                 No   

  assinatura_adicional_protecao_dispositivo  \
0                                        No   
1                                        No   
2                                       Yes   
3                                       Yes   
4                                        No   

  assinatura_adicional_suporte_tecnico_menos_tempo_espera assinatura_TV_cabo  \
0                                                Yes                     Yes   
1                                                 No                      No   
2                                                 No                      No   
3                                                 No                     Yes   
4                                                Yes                     Yes   

  assinatura_streaming_filmes        contrato fatura_online   forma_pagamento  \
0                          No        One year           Yes      Mailed check   
1                         Yes  Month-to-month            No      Mailed check   
2                          No  Month-to-month           Yes  Electronic check   
3                         Yes  Month-to-month           Yes  Electronic check   
4                          No  Month-to-month           Yes      Mailed check   

   gastos_mensais  gastos_diarios gasto_total  
0            65.6            2.19       593.3  
1            59.9            2.00       542.4  
2            73.9            2.46      280.85  
3            98.0            3.27     1237.85  
4            83.9            2.80       267.4

# Atualizar DataFrame

In [433]:
dados.to_csv('dataFrame.csv', index=False)